### ISSUE TITLE
Consider making ShaderModule compilation errors special

### ISSUE THREAD

**Issue by Kangz** [gpuweb/gpuweb#2119]
We'd like to be able to put most of the work involved by `createShaderModule` off-thread but this is awkward because of the way `createShaderModule` reports errors. We want to compile shader modules off-thread so that an application can create tons of shadermodule and async pipelines without stalling the rest of the WebGPU processing.

Currently errors during compilation of shader modules happen immediately during the `createShaderModule` call. This means that due to promise ordering of error scopes, either shader modules are created immediately, or (under the as-if rule) other device-timeline callbacks are delayed until we know the result of `createShaderModule`. What if we said that `createShaderModule` errors only appeared inside 1) the compilation messages or 2) when the shader module is used to create a pipeline?

FYI @Jiawei-Shao

**Comment by kvark** [comment]
What do you expect to get for compiling the shaders off-thread? Currently, the only work done during compilation is shader parsing and validation, which have very predictable performance, and should be fairly lightweight (~1ms per shader). And shaders are supposed to be created earlier in general (there is strictly less shaders than 2*pipelines). This is unlike pipeline creation, which can take arbitrarily more time due to driver work, and can be needed deeper in the frame.

**Comment by Kangz** [comment]
We expect to avoid stalls if 100+ shaders are compiled at once, and we expect to gain some compilation parallelism. (100ms stall is still important) Shader parsing and validation is also an area that will grow over time (see uniformity analysis or potential future with generics) but also some implementations might try to start doing backend compilation at this point (like the `MTLLibrary`).

**Comment by kvark** [comment]
It's just not totally clear to me why the shader module creation is where we draw the line. What if the user creates 100 buffers? May easily take 100ms as well. Does this mean buffer creation also needs to change the way errors are reported? Texture creation? etc. If you can suggest a good principle for drawing this line (and right now what we have is pretty strong), then the proposal will be much better.

**Comment by Kangz** [comment]
This is an interesting way to approach it. I think the reasoning here was that we see a need for pipelines to be created asynchronously but they aren't entirely asynchronous: layouts and shader modules that are part of them need to be created synchronously. Layouts are cheap but shader modules less so. However having consistent rules is very useful as well. Maybe we could just provide some helper library for the shadermodule stuff that will make sure only at least N compilation are in flight at a single time.

**Comment by dneto0** [comment]
> Currently, the only work done during compilation is shader parsing and validation, which have very predictable performance, and should be fairly lightweight (~1ms per shader).

I think this is a very temporary situation.  The technology trend is to do a lot more at shader creation time.

The Metal case is already pretty strongly emphasizing up-front shader compilation that can be quite heavyweight.
That is, we're taking #1064 very seriously. 

The buffers problem is qualitatively different, as I understand it. Allocating 100 buffers is expected to be slow. The driver architects push back on performance bug reports like that saying "allocate one giant buffer and suballocate out of that".

**Comment by kainino0x** [comment]
> Does this mean buffer creation also needs to change the way errors are reported?

FWIW it may be worth reconsidering the exact shape of error scopes for device-level work which isn't necessarily straight-line in principle. E.g. if I create a BGL, shader module, buffer, and command encoder in some order, why should each of them wait for the previous things? They're all independent. Some of them may not even hit the GPU at all (e.g. command encoders, in Dawn).

I don't have an alternate design in mind though. Haven't come up with any real ideas other than every operation reporting its own result (seems costly, maybe we can make it not be?). Maybe error sinks could be separate objects that can be optionally passed to certain operations to pull them out of the (device+thread)-global error sink?

**Comment by kvark** [comment]
>  The technology trend is to do a lot more at shader creation time.

Yes, we are trying to do much up-front at shader module creation. And also, we assume that people can re-use shader modules instead of creating new WGSL text for any small change they need. For users you want to help with this proposal, It would help to know what exactly forces them to create many shader modules on the spot (as opposed to - at initialization time).

> Allocating 100 buffers is expected to be slow

We may as well say that creating 100 shaders is expected to be slow.

---

If the problem can be stated is "shader module creation needs to be async like the pipeline creation", then why is the proposal not just `createShaderModuleAsync`? It confuses me that we are talking about a novel solution to something that we solved previously (for pipeline creation).

> FWIW it may be worth reconsidering the exact shape of error scopes for device-level work which isn't necessarily straight-line in principle.

We could generalize asynchronous resource creation by something like this:
```js
let factory = device.fork(); // resource creation in a fork may happen independently of the device, on a separate timeline
let module = factory.createShaderModule();
let pipeline = factory.createRenderPipeline();
factory.onCompletion(|| ...); // all resources created here are not ready until this promise is resolved
```

This would mean we can remove `createXxxPipelinesAsync` entirely, and users can create more than just pipelines asynchroniously, but also shader modules, buffers, bundles, etc.

**Comment by Kangz** [comment]
This specific suggestion was coming from experience with TF.js between others where they immediately try to create a graph of pipelines to execute ML inference. `createShaderModuleAsync` seems like a good way to solve it.

The factory proposal is interesting but it's a pretty big change and it would be nice that handles aren't available until the creation is finished so that we don't need to add validation checks for not-yet-ready everywhere. One the plus side, making objects available only once `onCompletion`'s callback is called is nice to avoid data races between the browser and the application.

**Comment by kvark** [comment]
I think the factory approach is superior to `createShaderModuleAsync` because:
  - it supports more than just shaders and pipelines, naturally
  - it gives users a better understanding of the timelines happening on GPU process side. I.e. "if I do 10 `createXxxPipelineAsync`, is this going to be parallel or sequential?"
  - it avoids round-trips when doing `createShaderModuleAsync` followed by `createXxPipelineAsync` (control flow has to go back to the content side before pipeline creation even has to be requested)

**Comment by Kangz** [comment]
As mentioned on chat the factory idea is interesting but it opens many complexities:

 - Can you create object using objects from the same factory? Another factory?
 - Do we add validation everywhere for not-yet-ready?
 - What about GetBindGroupLayout?
 - How redundant is it for when WebGPU will be allowed to work concurrently on multiple workers?

**Comment by magcius** [comment]
So, thoughts from working with this:

* Asynchronous compilation is a very, very big pain to manage. There are shaders that I pretty much need immediately, and the way my renderer is written (and most renderers, honestly), it's difficult to delay rendering until everything necessary is ready. Think about important shaders, like compute buffer copy helpers, depth resolves, antialiasing passes. Regardless of what objects are in your scene, you probably want these ready to go. Adding the back-pressure necessary to stall the rest of the loop is difficult.

* Additionally, knowing up-front what needs to be asynchronous and what doesn't is similarly difficult; the GL shader compilation model is currently not great but I like that it has the flexibility to switch an asynchronous task to a synchronous task without re-submitting the task, and Promises just sort of really make me uncomfortable in the context of a main loop ( see also: https://github.com/gpuweb/gpuweb/discussions/2085 )

I'd ask for a new way of designing an asynchronous API that gives me back a token that I can convert into synchronous compilation, but I feel like TAG isn't going to like that, so welp. :/

The factory approach seems like an OK idea, only I'd like an ability to block on the results from the GPU thread.

In any case, I expect that a hypothetical createShaderModuleAsync and createRenderPipelineAsync are going to be used together, and createShaderModule / createRenderPipeline are going to be used together.

> I think this is a very temporary situation. The technology trend is to do a lot more at shader creation time.

Weird; I actually see this going the other way. Individual shader modules don't have the necessary context to do full-pipeline optimization, so I would expect a lot of it is deferred to PSO creation time, but individual drivers will vary on the amount of work here. It's unfortunate that we often don't have a good view of what's expected to be expensive, and what's not.

**Comment by Kangz** [comment]
> I'd ask for a new way of designing an asynchronous API that gives me back a token that I can convert into synchronous compilation, but I feel like TAG isn't going to like that, so welp. :/

This is basically what the original suggestion was: `GPUShaderModule` becomes that token that by default can be compiled asynchronously but if it is used in a synchronous operation it is then either blocked on, or "stolen" to become a synchronous operations.

**Comment by magcius** [comment]
Considering I think the only thing you can do with a GPUShaderModule (currently) is pass it to createRenderPipeline, that's OK to me.

However, this limits us from attaching more functionality go GPUShaderModule in the future, like reflection data.

**Comment by litherum** [comment]
+1 for this request.

Can't we instead solve this by making `createShaderModule()` return a promise instead?

**Comment by magcius** [comment]
See my comment above for all the ergonomics issues with that; I think we still need a synchronous path.

**Comment by kainino0x** [comment]
> However, this limits us from attaching more functionality to GPUShaderModule in the future, like reflection data.

I think `compilationInfo` is the same, and that's a Promise right now. Unless we add a way for _JS_ to block on that, reflection could be exposed as a Promise as well. But maybe there's some way to bundle those two promises together.

**Comment by Kangz** [comment]
Reflection data is `compilationInfo` IMHO and could just be added there.

**Comment by greggman** [comment]
I want to make sure I understand this. I'm reading the spec [here https://www.w3.org/TR/webgpu/](https://www.w3.org/TR/webgpu/).

It seems to me according to that spec `createShaderModule` is already async, it just that if you actually use the shader then your app will stall until the shader is ready. I'm probably thinking in WebGL terms but getting the compilation info would seem like a signal that the shader is ready. So, if I wanted async compilation it seems like I'd write something like this

```
const shaders = {
   phong: { code: '', ...},
   postProcessing: { code, '', },
   ...
};

// compile all the shaders
for (const shader of shaders) {
  const module = device.createShaderModule({code: shader.code});
  shader.module = module;
  module.compilationInfo().then(info => { shader.ready = true; }).catch(e => console.error(e));
 }


function render() {
   if (shaders.phong.ready) {
      // draw stuff with shaders.phong.module)
   }
   if (shaders.postProcessing.ready) {
     // draw stuff with shaders.postProcessing.module;
   }
   requestAnimationFrame(render);
}
requestAnimationFrame(render);
```

would that not work?

**Comment by kainino0x** [comment]
It is async, I think the problem is that it also reports errors into the current error scope, which means it blocks `validation` and `out-of-memory` errors from reaching the user. TBH though, we should expect users to be putting error scopes only around the parts they care about, so maybe that's not a problem we need to solve.

(aside: in your example you'd also need to create a pipeline from the shader module before using it to render, but that's fine, it could be a createRenderPipelineAsync chained in the `.then`.)

**Comment by kvark** [comment]
@kainino0x 
> If an application didn't check those, it would simply not see any errors until later when the pipeline is used.

Isn't this a general problem with the error model? Why are we trying to poke holes in the model for pipelines and shader modules. If there is a general problem, it may need a general solution. e.g. if I'm creating a buffer incorrectly, and I'm not checking for it, I'm also only going to see the error later when the buffer is used.

**Comment by kainino0x** [comment]
Editors to figure out if we can come up with a solution here.

**Comment by kainino0x** [comment]
> What if we said that `createShaderModule` errors only appeared inside 1) the compilation messages or 2) when the shader module is used to create a pipeline?

The more I think on this the more I think this is a perfectly reasonable solution. It's perhaps a minor bump for CTS because we can't automatically catch pipeline 

Now, that's still making createShaderModule special. I think that's fine. But if we don't want to, then I think these are perfectly good alternatives:
- implementations just have to delay all error scope resolution until they've finished any asynchronous processing in createShaderModule (or other methods)
- get rid of error scope resolution ordering, or make it looser

**Comment by kainino0x** [comment]
> * implementations just have to delay all error scope resolution until they've finished any asynchronous processing in createShaderModule (or other methods)

That is, I think it's OK from a spec perspective. It compounds the difficulty of implementing concurrent shader module creation. But in principle it's okay if optimizations aren't necessarily easy to implement.

> * get rid of error scope resolution ordering, or make it looser

On second thought I don't think this makes a real difference in the implementation complexity. Still have to be able to delay the error scope containing the createShaderModule.

In the totally-ordered model of operations on the device timeline, concurrent work with fully-ordered error scopes could be modeled with two "instruction pointers": one for validation which has started and one for validation which has completed.

```
< is "validation has finished to this point"
> is "validation has started to this point"
>| is "validation is blocked at this point"

~ is "validation for this is in progress"
! is "validation for this is done but unreported"
. is "validation for this is done and reported"

  createSM  createBuffer  createPipeline  writeBuffer  submit
                          (using SM)                   (using SM+buffer)
<>
<     ~    >
<     ~          ~       >|  pipeline blocked on SM
<     ~          !       >|
<     !          !       >|
      .     <    !       >   pipeline unblocked
      .     <    !              ~        >
      .          .      <       ~        >
      .          .      <       ~              ~      >|  submit blocked on pipeline and buffer
      .          .      <       ~              !      >|
      .          .              .       <      !      >|  submit blocked on buffer
      .          .              .              .     <>   submit unblocked
      .          .              .              .     <    ~   >
      .          .              .              .          .  <>
```

With multithreading, this still models as one canonical device timeline, but where errors get reported *to* differs based on where the operation originated. (This seems fine.) Implementations don't have to actually single-thread the validation; they just have to do synchronization so it's not observable.

**Comment by toji** [comment]
Getting off topic now, but I see that Github just added [Mermaid](https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/) diagram support as well.

```mermaid
sequenceDiagram
    participant createSM 
    participant createBuffer
    participant createPipeline
    participant writeBuffer
    participant submit
    createSM->>createBuffer: 
    createBuffer->>createPipeline: 
    createPipeline->>createPipeline: pipeline blocked on SM
    createSM-->>createSM: validation completed
    createPipeline->>writeBuffer: pipeline unblocked
    createSM-->>createBuffer: validation completed
    writeBuffer->>submit:  
    submit->>submit:  submit blocked on pipeline and buffer
    createSM-->>createPipeline: validation completed
    submit->>submit:  submit blocked on buffer
     createSM-->>writeBuffer: validation completed
    submit->>submit:  submit unblocked
```

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/gpu-web-2022-06-29#consider-making-shadermodule-compilation-errors-special-2119">WGSL meeting minutes 2022-06-29</a></summary>

* KG: do we want these as part of ErrorScopes? Little weird - happen on different timelines. Some discussion about understanding these happen on different timelines, but handle like "await Promise.all" even if things happen out-of-order.
* KN: last week - moving toward agreement on my proposal - we make it look as-if everything happens in order. Internally, deal with reordering/threading. Not make it special.
* MM: that's what I was going to propose.
* KG: resolved.
* KN: Corentin was the most invested in this.
* KR: I was second-most-invested. I think this will cause more problems in the API. This has proven problematic in WebGL - all result reporting had to wait for shader compilation to complete.
* KN: errors are delayed - suppose there are situations where apps want to deal with allocation errors. Validation errors - don't think programmatically handling those. Another issue with separating out validation errors. Think it's OK if they're delayed.
* KG: worth considering - we do have ErrorScopes. You could contract the error scope, not include shader module compilation. Want to know about OOMs - tightly wrap those with their own ErrorScope. Don't have the problem you do in GL because we have ErrorScopes - think you need to just take care.
* MM: sounds like the behavior if we didn't do the fixup - start ErrorScope, compile, do something erroneous, ask for the error - proposal from Ken is that the second thing erroneous should make the error happen. What if the second thing is correct?
* KR: it will complete.
* Discussion about this. Main proposal - these aren't strictly ordered with respect to other errors in that ErrorScope.
* Discussion about behavior for correct vs. incorrect programs.
* KG:** if you want things to end up differently we need an actual proposal for this**.

</details>

**Comment by toji** [comment]
Kai and I were talking about this in today's editors meeting and were mildly surprised to find that we didn't link to the proposal that was made for this:

https://hackmd.io/s_QipO3NSiqQ_Ul7t2KbQQ

I think this has been linked elsewhere before, but it definitely should be listed here as well.
